crates/ruff_linter/src/linter.rs

```diff
@@ -329,6 +326,14 @@ pub fn check_path(
         }
     }
 
+    // Apply range suppressions before noqa, so that #noqa on a line already covered by a range
+    // will be reported as unused noqa rather than unused range.
+    // TODO: check if enabled?
+    {
+        let suppressions = Suppressions::from_tokens(locator.contents(), tokens);
+        suppressions.filter_diagnostics(&mut context);
+    }
```

Comment:
I think we should move the logic into `check_noqa` and rename the method to `check_suppressions`. This should also avoid having to iterate over all diagnostics twice, which can be expensive if there are many